### ISSUE TITLE
Support for JSONPath

### ISSUE THREAD

**Issue by TomasTokaMrazek** [mockoon/mockoon#995]
Object-path is very limited. It is not possible to select key name or traverse keyspace with wildcard patterns. JSONPath not only does support all of those but also can filter results and apply basic functions. This would greatly enhance rules capabilities without implementing scripting like Postman and SoapUI does.

This feature would also complement with xml-js in non-compact JSON mode for easier traversing through XML.

To showcase my use:

Given a SOAP service there's only one URL endpoint. Resource is selected by optional `SOAPAction` or by the request body itself. To do proper routing and mock SOAP service, parsing the body is a necessity.

Request for `GetUserCode`:
```
{
    "soap:Envelope": {
        "_attributes": {
            "xmlns:soap": "http://www.w3.org/2003/05/soap-envelope"
        },
        "soap:Body": {
            "id:GetUserCode": {
                "_attributes": {
                    "xmlns:id": "https://somedomain.com/id"
                },
                "id:Metadata": {
                    "id:Uuid": {
                        "_text": "123456789"
                    },
                    "id:Application": {
                        "_text": "mock"
                    }
                },
                "id:UserId": {
                    "_text": "1"
                }
            }
        }
    }
}
```

Request for `GetUserId`:
```
{
    "soap:Envelope": {
        "_attributes": {
            "xmlns:soap": "http://www.w3.org/2003/05/soap-envelope"
        },
        "soap:Body": {
            "id:GetUserId": {
                "_attributes": {
                    "xmlns:id": "https://somedomain.com/id"
                },
                "id:Metadata": {
                    "id:Uuid": {
                        "_text": "987654321"
                    },
                    "id:Application": {
                        "_text": "mock"
                    }
                },
                "id:UserCode": {
                    "_text": "A"
                }
            }
        }
    }
}
```

For every request there has to be different response. Right now there needs to be hacky solution trough full body and very generic regex. With JSONPath it could be done with path `$..soap:Body.*~` which results in:
```
[
  "id:GetUserCode"
]
```

With non-compact mode, where the JSON schema is always predictable with no dynamic keys, advanced filtering can be applied.

Non-compact request for 'GetUserId`:
```
{
    "elements": [
        {
            "type": "element",
            "name": "soap:Envelope",
            "attributes": {
                "xmlns:soap": "http://www.w3.org/2003/05/soap-envelope"
            },
            "elements": [
                {
                    "type": "element",
                    "name": "soap:Body",
                    "elements": [
                        {
                            "type": "element",
                            "name": "id:GetUserId",
                            "attributes": {
                                "xmlns:id": "https://somedomain.com/id"
                            },
                            "elements": [
                                {
                                    "type": "element",
                                    "name": "id:Metadata",
                                    "elements": [
                                        {
                                            "type": "element",
                                            "name": "id:Uuid",
                                            "elements": [
                                                {
                                                    "type": "text",
                                                    "text": "987654321"
                                                }
                                            ]
                                        },
                                        {
                                            "type": "element",
                                            "name": "id:Application",
                                            "elements": [
                                                {
                                                    "type": "text",
                                                    "text": "mock"
                                                }
                                            ]
                                        }
                                    ]
                                },
                                {
                                    "type": "element",
                                    "name": "id:UserCode",
                                    "elements": [
                                        {
                                            "type": "text",
                                            "text": "A"
                                        }
                                    ]
                                }
                            ]
                        }
                    ]
                }
            ]
        }
    ]
}
```

With given JSONPath `$..elements[?(@.type == "element" && @.name == "soap:Body")].elements[*].name` I get the same result, but with more precise control. XML element name prefixes can be arbitrary or even ommited. But with JSONPath, we can select value like this: `$..elements[?(@.type == 'element' && @.name.match(/UserCode/))].elements[*].text` with result:
```
[
  "A"
]
```
Those results are much more manageable and precise in response rules. 

There are only a few JSONPath libraries for JS with https://www.npmjs.com/package/jsonpath-plus being the most recent and feature complete, while no longer in active development.

**Comment by ajatkj** [comment]
Hi @255kb, I am picking this up. Will raise a PR when done.

**Comment by loganmzz** [comment]
Seen this merge, but online documentation doesn't seems to refer to JSON path. Have you some release plan ?

**Comment by 255kb** [comment]
This will be in the next release (v5.0.0) that is due around end of this week or early next week.

**Comment by 255kb** [comment]
📦 This is now available with [release v5.0.0](https://mockoon.com/releases/5.0.0/).
